### PR TITLE
Add massFractions as an optional argument to EOS routines

### PR DESCRIPTION
**Pull request by psharda**
This is the first part of using massFractions in EOS routines. This PR will add massFractions (given by the number of MassScalars) as an optional input to the EOS routines. Actual usage of the massFractions will follow in a separate PR.

### CODE REVIEW

File: src/EOS.hpp [hunk 1]
```diff
@@ -28,10 +30,15 @@ template <typename problem_t> struct EOS_Traits {
...
 template <typename problem_t> class EOS
 {
+
       public:
-	[[nodiscard]] AMREX_FORCE_INLINE AMREX_GPU_HOST_DEVICE static auto ComputeTgasFromEint(amrex::Real rho, amrex::Real Eint) -> amrex::Real;
-	[[nodiscard]] AMREX_FORCE_INLINE AMREX_GPU_HOST_DEVICE static auto ComputeEintFromTgas(amrex::Real rho, amrex::Real Tgas) -> amrex::Real;
-	[[nodiscard]] AMREX_FORCE_INLINE AMREX_GPU_HOST_DEVICE static auto ComputeEintTempDerivative(amrex::Real rho, amrex::Real Tgas) -> amrex::Real;
+	static constexpr int nmscalars_ = Physics_Traits<problem_t>::numMassScalars;
+	[[nodiscard]] AMREX_FORCE_INLINE AMREX_GPU_HOST_DEVICE static auto
+	ComputeTgasFromEint(amrex::Real rho, amrex::Real Eint, std::optional<amrex::GpuArray<amrex::Real, nmscalars_>> massFractions = {}) -> amrex::Real;
```
[the remark this answers is not in the document]
Comment (psharda) [reply src/EOS.hpp:38]: @BenWibking does this warning make sense?

[the remark this answers is not in the document]
Comment (BenWibking) [reply src/EOS.hpp:38]: I think this means there's a header `#include` missing.

[the remark this answers is not in the document]
Comment (psharda) [reply src/EOS.hpp:38]: GpuArray is defined in `AMReX_Array.H`, which is not directly included in EOS.hpp, but is accessible to it. I will use `#include AMReX_Array.H` and see if this warning goes away.

**Comment by psharda** [comment]
@BenWibking  The GPU build fails because the RadBeam test file `src/RadBeam/test_radiation_beam.cpp` uses `ComputeEintFromTgas` on line 221, without declaring it as a template function beforehand, like other tests do (e.g., RadMarshak). Can I add a template declaration for `ComputeEintFromTgas` with the optional argument in `test_radiation_beam.cpp`, somewhere after line 61, just like I did in the other tests? I am not sure what will the template function return though. In other test, it returns `return (alpha_SuOlson / 4.0) * std::pow(Tgas, 4);`

**Comment by BenWibking** [comment]
> @BenWibking The GPU build fails because the RadBeam test file `src/RadBeam/test_radiation_beam.cpp` uses `ComputeEintFromTgas` on line 221, without declaring it as a template function beforehand, like other tests do (e.g., RadMarshak). Can I add a template declaration for `ComputeEintFromTgas` with the optional argument in `test_radiation_beam.cpp`, somewhere after line 61, just like I did in the other tests? I am not sure what will the template function return though. In other test, it returns `return (alpha_SuOlson / 4.0) * std::pow(Tgas, 4);`

No, I think it's failing because:
```
../src/EOS.hpp(36): error: class "Physics_Traits<BeamProblem>" has no member "numMassScalars"
```
Log here: https://dev.azure.com/quokka-code/quokka-CI/_build/results?buildId=4372&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=d735d76d-5905-5d5d-cb1c-7198326fbfa2&l=202

**Comment by psharda** [comment]
ugh this branch became a mess. might need to close it and open a new PR

**Pull request closed without merging**